feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,188 @@
+
+	"magma/feg/gateway/sbi"
+	n7_client "magma/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl"
+	"magma/feg/gateway/services/n7_n40_proxy/n7"
+	"magma/feg/gateway/services/session_proxy/relay"
+	"magma/gateway/service_registry"
+	"magma/lte/cloud/go/protos"
+)
+
+const (
+	NOTIFY_PATH      = "/"
+	UnableToDeliver  = 3002
+	EncodedSessionId = "encodedSessionId"
+	ShutdownTimeout  = 10 * time.Second
+)
+
+type NotificationHandler struct {
+	Config        *n7.N7ClientConfig
+	NotifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+}
+
+func NewNotificationHandler(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) (*NotificationHandler, error) {
+	handler := &NotificationHandler{
+		Config:        config,
+		NotifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+	urlDef, err := url.ParseRequestURI(config.NotifyApiRoot)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.NotifyServer.POST(basePath+"/update", handler.PostSmPolicyUpdateNotification)
+	handler.NotifyServer.POST(basePath+"/terminate", handler.PostSmPolicyTerminateNotification)
+
+	return handler, nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.Config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.NotifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.NotifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.NotifyServer.Shutdown(ctx)
+}
+
+func (handler *NotificationHandler) PostSmPolicyUpdateNotification(ctx echo.Context) error {
+	client, err := relay.GetSessionProxyResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	defer client.Close()
+
+	var smUpdateNotify n7_client.SmPolicyNotification
+	err = ctx.Bind(&smUpdateNotify)
+	if err != nil {
+		glog.Errorf("Invalid SmPolicyNotification received: %s", err)
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+	smPolicyDecision := smUpdateNotify.SmPolicyDecision
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for UpdateNotify- %s", err)
+		return err
+	}
+
+	policyReauthProto := n7.GetPolicyReauthRequestProto(sessionId, imsi, smPolicyDecision)
+	ans, err := client.PolicyReAuth(context.Background(), policyReauthProto)
+	if err != nil {
+		glog.Errorf("Error relaying N7 policy reauth request to gateway: %s", err)
+		return err
+	}
+	partialSuccessReport := n7.GetPartialSuccessReportN7(ans)
+	if partialSuccessReport == nil {
+		return ctx.NoContent(http.StatusOK)
+	}
```

Comment:
this is too convoluted.  `GetPartialSuccessReportN7` does to many thing. The return is also doing to many things, not only provides the value but serves as an error. 

Why don't you create a function in this file to process the response from `PolicyReAuth` which will return eitner `ctx.JSON(http.StatusOK, partialSuccessReport)` or `ctx.NoContent(http.StatusOK)`.  You can put part of the logic of `GetPartialSuccessReportN7` in this function maybe.

Note in go it is better to return error inside the iff, and return the proper case outside the iff (here is done reversed)

Reply:
IN CONVERSIONS.GO
```
// BuildPartialSuccessReportN7 define this function builds a partial success repoirt. 
// Highlight that this function will generate a report no matter if the answer was really partial
func BuildPartialSuccessReportN7(failedRules *protos.Rules (???)) *n7_sbi.PartialSuccessReport {
	failedRuleReports := []n7_sbi.RuleReport{}
	for ruleId, failureCode := range failedRules {
		failedRuleReport := n7_sbi.RuleReport{
			PccRuleIds:  []string{ruleId},
			RuleStatus:  n7_sbi.RuleStatusINACTIVE,
			FailureCode: getFailureCodeN7(failureCode),
		}
		failedRuleReports = append(ruleReports, failedRuleReport)
	}

	return &n7_sbi.PartialSuccessReport{
		FailureCause: n7_sbi.FailureCausePCCRULEEVENT,
		RuleReports:  &failedRuleReports,
	}
}
```

IN NOTIFY_HANDLER 
```
    	if isItPartialSuccess(ans) {
		ctx.JSON(http.StatusOK,, n7.GetPartialSuccessReportN7(ans))
	}

	return ctx.NoContent(http.StatusOK)
}

//(maybe that could also go in CONVERSIONS.GO)
// isItPartialSuccess define this clearly
func isItPartialSuccess(raa *protos.PolicyReAuthAnswer) bool; {
	if raa.Result == protos.ReAuthResult_UPDATE_INITIATED ||
		raa.Result == protos.ReAuthResult_UPDATE_NOT_NEEDED ||
		len(raa.FailedRules) == 0 {
		// no errors to send
		return true
	}
	return false 
}


```



Maybe we should start thinking on a `conversion.go` file under sbi 
See how we extend swagger auto generated objects [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fcloud%2Fgo%2Fservices%2Ffeg%2Fobsidian%2Fmodels%2Fconversion.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=38&start_col=0&end_row=38&end_col=0). Sme we do with some autogenerated protos [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fcloud%2Fgo%2Fprotos%2Fmock_core_conversions.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=13&start_col=10&end_row=13&end_col=10)

Reply:
Lets do the change as it is but I think adding `isItPartialSuccess` to a  `conversions.go` may be the best . Do you know if this is partial logic is used somewhere else

Reply:
i dont understand the logic: when there is a partial success we modify the contex? but still return ok?